R/generate_dust_equation.R

```diff
@@ -92,6 +93,19 @@ generate_dust_equation_alloc <- function(eq, data_info, dat, rewrite, gpu) {
 }
 
 
+generate_dust_equation_copy <- function(eq, data_info, dat, rewrite, gpu) {
+  x <- dat$data$output$contents[[data_info$name]]
+  if (data_info$rank == 0L) {
+    sprintf("output[%s] = %s;", rewrite(x$offset), rewrite(eq$lhs))
+  } else {
+    c(sprintf("for (size_t i = 0; i < %s; ++i) {",
+              rewrite(data_info$dimnames$length)),
+      sprintf("  output[%s + i] = %s[i];", rewrite(x$offset), rewrite(eq$lhs)),
+      "}")
```

Comment:
An alternative way here would be to use `copy_n` but I am not sure if this is much more readable? (untested!)

```suggestion
    sprintf("std::copy_n(%s.begin(), %s, %s.begin() + %s);", rewrite(eq$lhs), rewrite(data_info$dimnames$length), dat$meta$output, rewrite(x$offset))
```

If you could use `dat$meta$output` rather than the string literal `"output"` too that'd be great, though I can't imagine we'll ever change that now...